### PR TITLE
new case: Create nwfilter-binding when aready present

### PR DESCRIPTION
**Pull request by juteman**
Senario 1: start vm with nwfilterred defined in xml
Senario 2: start vm with nwfilterred do not defined in xml

Signed-off-by: qiulzhan <qiulzhan@redhat.com>

### CODE REVIEW

**Comment by juteman** [comment]
```
JOB ID     : 34bf54a5de19509b2d40d3469d7317c73a474478
JOB LOG    : /root/avocado/job-results/job-2018-08-23T17.02-34bf54a/job.log
 (1/1) type_specific.io-github-autotest-libvirt.filter_aready_present_binding.filter_not_binding: PASS (14.16 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 26.96 s
```

**Comment by juteman** [comment]
```
JOB ID     : 67523ee054d5ce20edae32bfc4589a61b1ce30d9
JOB LOG    : /root/avocado/job-results/job-2018-08-23T17.03-67523ee/job.log
 (1/1) type_specific.io-github-autotest-libvirt.filter_aready_present_binding.filter_is_binding: PASS (11.10 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 23.77 s

```

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    def set_env(is_nwfilter_define):
+        vmxml = libvirt_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+        iface_xml = vmxml.get_devices('interface')[0]
+        vmxml.del_device(iface_xml)
+        new_iface = interface.Interface('network')
+        new_iface.xml = iface_xml.xml
+        new_iface.type_name = "network"
+        iface_target = {'dev': "vnet0"}
+        new_iface.target = iface_target
+        source = {'network': "default", 'bridge': "virbr0"}
```
Comment (kylazhang) [inline libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: please move all those hard code in cfg

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            expected_match = params.get("expected_match")
+            cmd_restart = params.get("cmd_restart")
+            filter_binding_name = params.get("filter_binding_copy")
+
+            ret = virsh.nwfilter_binding_list(debug=True)
+            utlv.check_cmd_expected(check_cmd, expected_match, True)
+            cmd_res = process.run(cmd_restart, shell=True)
+            if cmd_res.exit_status:
+                test.fail("fail to restart libvirtd")
+            ret = virsh.nwfilter_binding_list(debug=True)
```
Comment (kylazhang) [inline libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: as above

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            filterrefs_dict = {}
+            filterrefs_dict['name'] = filter_binding_name
+            filterrefs_dict['parameters'] = [
+                {'name': "MAC", 'value': new_iface.mac_address}]
+            binding.filterref = binding.new_filterref(**filterrefs_dict)
+            logging.debug("binding is %s" % new_binding)
+            ret_create = virsh.nwfilter_binding_create(new_binding.xml,
+                                                       debug=True)
+            utlv.check_result(ret_create, failed_msg)
+
```
Comment (kylazhang) [inline libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: I saw original case have detach-interface operation, you do not do it?

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        elif not is_nwfilter_define:
+            # get params for senario2
+            check_cmd = params.get("check_cmd")
+            expected_match = params.get("expected_match")
+            cmd_restart = params.get("cmd_restart")
+            filter_binding_name = params.get("filter_binding_copy")
+
+            ret = virsh.nwfilter_binding_list(debug=True)
+            utlv.check_cmd_expected(check_cmd, expected_match, True)
+            cmd_res = process.run(cmd_restart, shell=True)
```
Comment (kylazhang) [inline libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: Use utils_libvirtd module

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    vm = env.get_vm(vm_name)
+    status_error = "yes" == params.get("status_error")
+    vmxml_backup = libvirt_xml.vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    filter_name = params.get("filter_name")
+    is_nwfilter_define = "yes" == params.get("is_nwfilter_define")
+    vnet0_binding = os.path.join(data_dir.get_tmp_dir(), "vnet0_binding.xml")
+    filter_binding_name = params.get("filter_binding_name")
+    failed_msg = params.get("expected_failed")
+
+    def set_env(is_nwfilter_define):
```
[the remark this answers is not in the document]
Comment (juteman) [reply libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: it my miss. forgot to delete the param of func after code change.
I fix it and add command of that

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        filterrefs_dict['name'] = filter_binding_name
+        filterrefs_dict['parameters'] = [
+            {'name': "MAC", 'value': new_iface.mac_address}]
+        binding.filterref = binding.new_filterref(**filterrefs_dict)
+        logging.debug("binding is %s" % binding)
+        # list filter
+        if not is_nwfilter_define:
+            virsh.nwfilter_binding_delete(new_iface.target['dev'], debug=True)
+
+        ret = virsh.nwfilter_binding_list(debug=True)
```
[the remark this answers is not in the document]
Comment (juteman) [reply libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: Add func to check expected now

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py (hunk 1 above)
Comment (juteman) [reply libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: add this code to cfg

File: libvirt/tests/src/nwfilter/filter_aready_present_binding.py (hunk 3 above)
Comment (juteman) [reply libvirt/tests/src/nwfilter/filter_aready_present_binding.py]: this case have no detach-interface

**Pull request merged**